Project autocompleter with hierarchy

b12f opened the pull request:
- [x] Select parent project on create new project
- [x] Select parent project on copying a project
- [x] Change the parent project information on the project settings
- [x] Change project of a work package on split screen
- [x] Change project of a work package on full view
- [x] Change project of a work package on work package list column
- [x] Project filter on the boards (as include project is not implemented yet)
- [x] Create a new work package from outside of any project
- [x] Notification settings
- [x] Invite user modal
- [x] Administration > Users > Specific User > Projects: New project selection
- [x] Administration > Placeholder users > Specific placeholder users > Projects: New project selection
- [x] Administration > Groups > Specific groups > Projects: New project selection
- [x] Administration > Work Packages > Custom actions > New custom action
- [x] Add a project filter into a new cost report


Closes https://community.openproject.org/projects/openproject/work_packages/40286/activity

Commit:
Create initial project completer

Creates an initial version of the project-autocompleter and does a bit of refactoring on the user-autocompleter to keep
the two in line. Also changes some imports of modules around to prevent circular dependencies.

Commit:
Load projects via signalling in project-autocompleter

This changes the projects loading logic aroundto use signalling. After
loading the list of projects, the data gets transformed into the same
tree structure that's in use in project-include.

Afterwards, we flatten the tree back to a list so that ng-select can
consume it. The hierarchy is "faked" with paddings.

HDinger reviewed (commented):
The approach is fine for me, but since I was part of the development I'd say we let @oliverguenther have a look as well.

oliverguenther commented:
Can we internally make use of `op-autocompleter` again as the component we wrap, and pass in templates like we do in other places? https://github.com/opf/openproject/blob/dev/frontend/src/app/features/user-preferences/notifications-settings/inline-create/notification-setting-inline-create.component.html#L23

This avoids us applying fixes already fixed in op-autocompleter, such as initially triggering the typeahead to search and repositioning during opening the dropdown.

Commit:
Initial IUM project select implementation

The project selection in the invite user modal has some interesting
problems regarding filtering that it needs to solve, so it'll be the
first usecase to be implemented.

Commit:
Building of data structure for project autocompleter seems to work

Commit:
op-autocompleter appears broken

Commit:
Change to getOptionsFn

Commit:
Insert project autocompleter in dyanmic forms

Commit:
Change mapResultsFn interfacing

Commit:
Improvements to dynamic form handling

Commit:
Use project autocompleter in edit fields

Commit:
Change DatasetInputs into a stateful function

The DatasetInputs decorator was extending the component class, causing
the Angular dependency injection system not to recognize the component
anymore. DatasetInputs is removed and changed into a function called
populateInputsFromDataset that can be called in the contructor of
the necessary component.

crohr committed:
Upgrade pullpreview instance

Commit:
Fix project autocompleter for filters

Commit:
Merge branch 'feature/40286-project-hierarchy-selector' of github.com:opf/openproject into feature/40286-project-hierarchy-selector

Commit:
Use project autocompleter for principal role selection

Commit:
Use project autocompleter for custom actions

Commit:
Fix cost reporting project filter select

Commit:
Fix APIFilters dataset inputs

oliverguenther committed:
Allow user_action filtered by move_work_packages

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/40286-project-hierarchy-selector

Commit:
Fix linting errors, have working build

Commit:
Fix CF filter select

Commit:
Fix failing spec

Commit:
Fix failing spec

Commit:
Fix failing spec

Commit:
Fix failing spec

Commit:
Fix failing spec

Commit:
Merge branch 'dev' into feature/40286-project-hierarchy-selector

Commit:
Fix broken merge

oliverguenther committed:
Fix formly specs after modifying selectInput

Commit:
Fix import of invite user button module

Commit:
Merge branch 'feature/40286-project-hierarchy-selector' of github.com:opf/openproject into feature/40286-project-hierarchy-selector

Commit:
Fix project input label and text input selection in custom actions spec

Commit:
Fix failing spec

Commit:
Fix failing spec

Commit:
Fix backend value given into the reporting project filter

Commit:
Fix bindLabel of op-autocompleter

Commit:
Add back disabled reason to notification setting inline create project selection

Commit:
Add not nil check to project cost reporting filter

Commit:
Fix selecting of project in bulk move wp spec

Commit:
Fix selecting of project in bulk move wp spec

Commit:
Fix move work package page reload on project select

Commit:
Fix group project spec failing because project autocompleter had zero width

Commit:
Fix update cost report spec by casting IDs to integers before checking for equality

Commit:
Wait a bit in flickering spec

oliverguenther committed:
Better waiting for the correct action to take place

Clicking move and follow might be broken due to the location.href
in the refresh-on-form-changes component

Commit:
Fix validator in invite user modal

Commit:
Merge branch 'feature/40286-project-hierarchy-selector' of github.com:opf/openproject into feature/40286-project-hierarchy-selector

Commit:
Fix cost reporting project switch spec

Commit:
Fix principal membership project select in spec

Commit:
Merge branch 'dev' into feature/40286-project-hierarchy-selector

HDinger reviewed (changes requested):
There are some things not working as specified:

- [x] The current search term is not highlight in the results
- [x] Changing the project of a work package on a work package list column is not working
- [x] Two of the project filters (on boards) find too many results. They should only find subprojects of the current project
  - [x] "including subprojects"
  - [x] "only subproject"
- [x] In the invite user modal, the dropdown is misplaced
- [x] All of the autocompleters in the "admin -> users" sections find too many results. Those projects that are already selected, should not be selectable again
  - [x] Specific User > Projects: New project selection
  - [x] Placeholder users > Specific placeholder users > Projects: New project selection
  - [x] Groups > Specific groups > Projects: New project selection
- [x] In Administration > Work Packages > Custom actions
  - [x] Changing the project in the preconditions is not saved
  - [x] In the action field "project", the old autocompleter is used
- [x] In cost reports
  - [x] adding multiple projects and saving that, results in a crash after reload
  - [x] Freshly adding a project filter, shows the old project field

HDinger committed:
Some eslint and rubocop optimizations (no change in functionality)

HDinger committed:
Move SearchHighlightDirective to AutocompleterModule in order to use it in the project autocompleter. The highlightDirective needed to be extended so that it can highlight dynamically and not only once at the first trigger.

HDinger committed:
Add API filter for membership autocompleters to avoid that already selected projects can be chosen again

HDinger committed:
Add ancestors filter to show only relevant subprojects in "only subprojects" and "including subprojects" filter

HDinger committed:
Move the project autocompleter for custom action to the autocomplete decoration component in order to keep the logic of adding hidden field values when chosing a different option. Otherwise the project selection would never be saved. Unfortunately, there is some mapping neede as the the project autocompleter and the decorator deal with different types.

HDinger committed:
Show the project autocompelter also when initially adding the project filter where `f_inst` is nil. Furhter let the filter be able to deal with multiple projects selected.

HDinger committed:
Avoid too often replacement of highlighted strings

HDinger committed:
Append the dropdown to the modal overlay to avoid offsets

HDinger committed:
Handle case that no projects have been pre-selected when adding a new membership

HDinger committed:
Load form if availableValues link is not loaded yet so that the project edit field always shows the correct values

oliverguenther committed:
Don't show archived projects in memberships

oliverguenther committed:
Remove outdated comment

oliverguenther committed:
Omit key value 3.1

HDinger committed:
Correctly parse both sides of the comparison to lower case

oliverguenther committed:
Wait until body is reloaded

There is a race condition between reloading the project and clicking
copy and follow.

The project results in a reload of the page, which is sometimes followed by
copying before the reload is done.
